x-pack/plugins/fleet/server/services/epm/archive/save_to_es.ts

```diff
@@ -0,0 +1,183 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { extname } from 'path';
+import { isBinaryFile } from 'isbinaryfile';
+import mime from 'mime-types';
+import {
+  PACKAGE_ASSETS_INDEX_NAME,
+  InstallablePackage,
+  InstallSource,
+  PackageAssetReference,
+} from '../../../../common';
+import { CallESAsCurrentUser } from '../../../types';
+import { appContextService } from '../../../services';
+import { getArchiveEntry } from './index';
+
+// could be anything, picked this from https://github.com/elastic/elastic-agent-client/issues/17
+const MAX_ES_ASSET_BYTES = 4 * 1024 * 1024;
+
+const packageAssetMappings = {
+  package_name: { type: 'keyword' },
+  package_version: { type: 'keyword' },
+  install_source: { type: 'keyword' },
+  asset_path: { type: 'keyword' },
+  media_type: { type: 'keyword' },
+  data_utf8: { type: 'text', index: false },
+  data_base64: { type: 'binary' },
+};
+
+export interface PackageAsset {
+  package_name: string;
+  package_version: string;
+  install_source: string;
+  asset_path: string;
+  media_type: string;
+  data_utf8: string;
+  data_base64: string;
+}
+
+export const ensurePackagesIndex = async (opts: { callCluster: CallESAsCurrentUser }) => {
+  const { callCluster } = opts;
+  const logger = appContextService.getLogger();
+  const indexExists = await callCluster('indices.exists', { index: PACKAGE_ASSETS_INDEX_NAME });
+  if (!indexExists) {
+    try {
+      const clientParams = {
+        index: PACKAGE_ASSETS_INDEX_NAME,
+        body: {
+          mappings: {
+            properties: packageAssetMappings,
+          },
+        },
+      };
+      await callCluster('indices.create', clientParams);
+    } catch (putErr) {
+      logger.error(`${PACKAGE_ASSETS_INDEX_NAME} could not be created`);
+    }
+  }
+};
+
+export const saveArchiveEntriesToES = async (opts: {
+  callCluster: CallESAsCurrentUser;
+  paths: string[];
+  packageInfo: InstallablePackage;
+  installSource: InstallSource;
+}) => {
+  const { callCluster, paths, packageInfo, installSource } = opts;
+  await ensurePackagesIndex({ callCluster });
+  const bulkBody = await createBulkBody({ paths, packageInfo, installSource });
+  const results: BulkResponse = await callCluster('bulk', { body: bulkBody });
+  return results;
+};
+
+export async function createBulkBody(opts: {
+  paths: string[];
+  packageInfo: InstallablePackage;
+  installSource: InstallSource;
+}) {
+  const { paths, packageInfo, installSource } = opts;
+  const bulkBody = await Promise.all(
```

Comment:
Just to be sure I get these correctly, for one package we are going to build a bulk request with all the files? Are you not concerned that this could be big and loading too many things in memory?

Reply:
Yes, you're reading that correctly. It's mainly that way because we're currently treating asset installation as all-or-nothing proposition and I because I believe it faster/safer to do fewer network connections.  The [bulk docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-bulk.html) say
"This reduces overhead and can greatly increase indexing speed."

I did just learn the [default max document size](https://www.elastic.co/guide/en/elasticsearch/current/general-recommendations.html#maximum-document-size) is 100Mb but we don't have any approaching that *yet*. Our largest packages aren't half that size. There's definitely a possibility of files larger than 100Mb eventually and I will add a point to the proposal discussion about it.